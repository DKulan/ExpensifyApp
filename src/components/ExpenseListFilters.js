import React from 'react'
import {connect} from 'react-redux'

import {DateRangePicker} from 'react-dates'
import {setStartDate, setEndDate, setTextFilter, sortByAmount, sortByDate} from '../actions/filters'


class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({
            calendarFocused
        }))
    }

    render() {
        const {dispatch, filters} = this.props

        return (
            (
                <div>
                    <input
                        type="text"
                        value={filters.text}
                        onChange={(e) => {
                            dispatch(setTextFilter(e.target.value))
                        }}
                    />
                    <select
                        value={filters.sortBy}
                        onChange={(e) => {
                            if (e.target.value === 'date') {
                                dispatch(sortByDate())
                            } else {
                                dispatch(sortByAmount())
                            }
                        }}>
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                    </select>
                    <DateRangePicker
                        startDate={filters.startDate}
                        endDate={filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        showClearDates={true}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                </div>
            )
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilters)