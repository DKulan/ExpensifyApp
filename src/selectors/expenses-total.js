const getExpensesTotal = (expenses) => {
  if (expenses.length === 0) {
    return 0
  } else {
    return expenses.map((expense) => expense.amount).reduce((total, amount) => {
      return total + amount
    })
  }
}

export {getExpensesTotal}