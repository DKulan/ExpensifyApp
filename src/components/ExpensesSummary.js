import React from 'react'
import {connect} from 'react-redux'
import selectExpenses from '../selectors/expenses'
import {getExpensesTotal} from '../selectors/expenses-total'
import numeral from 'numeral'


export class ExpensesSummary extends React.Component {
  render() {
    const {expenseCount, expenseTotal} = this.props

    return (
      <h2>Viewing {expenseCount} expense{(expenseCount === 0 || expenseCount > 1) && 's'} totalling {numeral(expenseTotal/100).format('$0,0.00')}</h2>
    )
  }
}

const mapStateToProps = (state) => ({
  expenseCount: selectExpenses(state.expenses, state.filters).length,
  expenseTotal: getExpensesTotal(selectExpenses(state.expenses, state.filters))
})

export default connect(mapStateToProps)(ExpensesSummary)