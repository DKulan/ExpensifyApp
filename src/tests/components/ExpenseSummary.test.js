import React from 'react'
import {shallow} from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'
import {getExpensesTotal} from '../../selectors/expenses-total'
import moment from 'moment'


const expenses = [{
  id: '1',
  description: 'gum',
  note: '',
  amount: 195,
  createdAt: 0
}, {
  id: '2',
  description: 'rent',
  note: '',
  amount: 109500,
  createdAt: moment(0).subtract(4, 'days').valueOf()
}, {
  id: '3',
  description: 'credit card',
  note: '',
  amount: 4500,
  createdAt: moment(0).add(4, 'days').valueOf()
}]

test('should correctly render ExpensesSummary with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235}/>)

  expect(wrapper).toMatchSnapshot()
})

test('should correctly render ExpensesSummary with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={500238}/>)

  expect(wrapper).toMatchSnapshot()
})