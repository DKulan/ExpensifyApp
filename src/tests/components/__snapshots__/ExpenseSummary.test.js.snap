// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<h1>
  Viewing 
  1
   expense
   totalling 
  $0.00
</h1>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<h1>
  Viewing 
  23
   expense
  s
   totalling 
  $0.00
</h1>
`;
